server/api_archive.go

```diff
@@ -0,0 +1,178 @@
+package server
+
+// This is a low-level archiver which zips and sends GCA server persistent file
+// data using a POST command. This is intended to be used by an external service
+// to provide archival snapshots of raw data in the system. Only public data is
+// provided, with private keys zeroed before sending.
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+// The contents of a README file which will be inserted into the archive.
+const ReadmeContents = `This archive contains uninterpreted server files.
+These files all contain publicly available information
+and additional work is needed to stand up a new server
+using them.
+`
+
+// Zip archive helper.
+type zipArchiveWriter struct {
+	buffer *bytes.Buffer
+	znw    *zip.Writer
+}
+
+// Create a new zip archive writer.
+func NewArchive() *zipArchiveWriter {
+	buf := new(bytes.Buffer)
+	return &zipArchiveWriter{
+		buffer: buf,
+		znw:    zip.NewWriter(buf),
+	}
+}
+
+// Close the archive writer and return a buffer to its contents.
+func (arc *zipArchiveWriter) Close() (*bytes.Buffer, error) {
+	if err := arc.znw.Close(); err != nil {
+		return nil, nil
+	}
+	return arc.buffer, nil
+}
+
+func (arc *zipArchiveWriter) AddFile(reader io.Reader, name string, modTime time.Time) error {
+	writer, err := arc.znw.CreateHeader(&zip.FileHeader{
+		Name:     name,
+		Method:   zip.Deflate,
+		Modified: modTime,
+	})
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(writer, reader); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ArchiveHandler provides the GET /archive api endpoint. It returns
+// uninterpreted file data intended to be used by an external service
+// to produce an archive.
+func (gcas *GCAServer) ArchiveHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET method is allowed", http.StatusMethodNotAllowed)
+		return
+	}
+	if r.ContentLength != 0 {
+		http.Error(w, "Request should not have a body", http.StatusBadRequest)
+		return
+	}
+	if !gcas.ApiArchiveRateLimiter.Allow() {
+		http.Error(w, fmt.Sprintf("Too many requests, this server allows %v every %v", apiArchiveLimit, apiArchiveRate), http.StatusTooManyRequests)
+		return
+	}
+	arc := NewArchive()
+
+	// File ordering matters here. We are not locking the files, and relying on the file writes being
+	// append only while happening in a single write. We must archive the files in reverse order from
+	// their dependency order.
+	for _, name := range PublicFiles {
```

Comment:
I wouldn't usually write this because it's mundane, but there's enough other comments here I figured I'd mention. Typically the variable name you'd use here for `name` is `pf`, because you are iterating through a list of public files, and so the `pf` name makes it really easy later on in the loop to reference the top and realize that the variable is referring to an element of the main list.

Though I can see why you picked the name `name` (short for filename), that context requires a bit more of a leap. It's not a big deal here because the loop is short so the code is still plenty readable, but if your loops get longer and more involved it helps to have more strict conventions around what everything gets named

Reply:
That's a good point. Generic names tend to make the code harder to understand in the long term.